### PR TITLE
[SobolEngine] Update direction numbers to 21201 dims

### PR DESCRIPTION
**Pull request by Balandat**
Performs the update that was suggested in https://github.com/pytorch/pytorch/issues/41489

Adjust the functionality to largely match that pf the scipy companion PR https://github.com/scipy/scipy/pull/10844/, including
- a new `draw_base2` method
- include zero as the first point in the (unscrambled) Sobol sequence

The scipy PR is also quite opinionated if the `draw` method doesn't get called with a base 2 number (for which the resulting sequence has nice properties, see the scipy PR for a comprehensive discussion of this). 

Note that this update is a **breaking change** in the sense that sequences generated with the same parameters after as before will not be identical! They will have the same (better, arguably) distributional properties, but calling the engine with the same seed will result in different numbers in the sequence.

Test Plan:
```
from torch.quasirandom import SobolEngine

sobol = SobolEngine(3)
sobol.draw(4)

sobol = SobolEngine(4, scramble=True)
sobol.draw(5)

sobol = SobolEngine(4, scramble=True)
sobol.draw_base2(2)
```

Differential Revision: D25657233

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
All Sobol related tests are failing, can you please look at those (and either update the tests or fix underlying failures)

**Comment by Balandat** [comment]
I didn't touch the tests yet b/c I wanted to get a read on how reasonable putting the constants into the humongous constant C-array. If that seems reasonable I'm happy to fix up the units.

**Comment by Balandat** [comment]
@malfet I updated the PR so that it implements the same behavior as the Sobol sampler in the scipy PR - the implementation and statistical properties were discussed at length with experts in the field there (443 comments on the PR and counting), so I am very confident about this being the correct thing to do. 

I adapted the tests from the scipy PR. I can add more of the old tests back in, but I will have to update target numbers b/c of the breaking change to the direction numbers and the inclusion of zero as the first point.

File: aten/src/ATen/native/SobolEngineOps.cpp [hunk 1]
```diff
@@ -126,19 +126,28 @@ Tensor& _sobol_engine_initialize_state_(Tensor& sobolstate, int64_t dimension) {
...
-  sobolstate.select(0, 0).fill_(1);
-
   /// Use a tensor accessor for `sobolstate`
   auto ss_a = sobolstate.accessor<int64_t, 2>();
-  for (int64_t d = 0; d < dimension; ++d) {
+
+  /// First row of `sobolstate` is all 1s
+  for (int64_t m = 0; m < MAXBIT; ++m) {
+    ss_a[0][m] = 1;
+  }
```
Comment (malfet) [inline aten/src/ATen/native/SobolEngineOps.cpp:135]: Nit
```suggestion
  std::fill=n(ss_a[0], MAXBIT, 1);
```

File: aten/src/ATen/native/SobolEngineOps.cpp [hunk 2]
```diff
@@ -126,19 +126,28 @@ Tensor& _sobol_engine_initialize_state_(Tensor& sobolstate, int64_t dimension) {
...
+
+  /// Remaining rows of sobolstate (row 2 through dim, indexed by [1:dim])
+  for (int64_t d = 1; d < dimension; ++d) {
     int64_t p = poly[d];
     int64_t m = bit_length(p) - 1;
 
+    // First m elements of row d comes from initsobolstate
     for (int64_t i = 0; i < m; ++i) {
       ss_a[d][i] = initsobolstate[d][i];
     }
```
Comment (malfet) [inline aten/src/ATen/native/SobolEngineOps.cpp:145]: Nit
```suggestion
    std::copy(ss_a[d], ssa_[d] + m, initsobolstate[d]);
```

File: test/test_torch.py [hunk 3]
```diff
@@ -1560,102 +1560,257 @@ def test_generator_cpu(self):
...
+        #     expected_3d = [0.133490, 0.480183, 0.855304, 0.970967, 0.345844]
+        #     engine_3d = torch.quasirandom.SobolEngine(3, scramble=True)
+        #     actual_3d = engine_3d.draw(5)
+        #     self.assertEqual(actual_3d[:, 0], expected_3d)
+
+        # def test_sobolengine_scrambled_highdim(self):
+        #     engine = torch.quasirandom.SobolEngine(1111, scramble=True)
+        #     draws = engine.draw(1000)
+        #     self.assertTrue(torch.all(draws <= 1))
+        #     self.assertTrue(torch.all(draws >= 0))
```
Comment (malfet) [inline test/test_torch.py:1813]: Please delete old test source
```suggestion

```

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/SobolEngineOps.cpp (hunk 2 above)
Comment (Balandat) [reply aten/src/ATen/native/SobolEngineOps.cpp]: I have very little understanding of the C++ internals of torch, but this fails (your suggestion above as well). Note that initsobolstate is a C array not a tensor, not sure if that's what's the issue here. I'll leave this and the `fill_n` above that also fails as is for now.
...

stderr: caffe2/aten/src/ATen/native/SobolEngineOps.cpp:141:32: error: invalid operands to binary expression ('TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' and 'int64_t' (aka 'long'))
    std::copy(ss_a[d], ss_a[d] + m, initsobolstate[d]);
                       ~~~~~~~ ^ ~
caffe2/aten/src/ATen/TensorOperators.h:110:22: note: candidate function not viable: no known conversion from 'TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' to 'const at::Tensor' for 1st argument
AT_FORALL_BINARY_OPS(DEFINE_OPERATOR)
                     ^
caffe2/aten/src/ATen/TensorOperators.h:82:33: note: expanded from macro 'AT_FORALL_BINARY_OPS'
#define AT_FORALL_BINARY_OPS(_) \
                                ^
caffe2/aten/src/ATen/TensorOperators.h:102:22: note: expanded from macro 'DEFINE_OPERATOR'
static inline Tensor operator op(const Tensor & x, Scalar y) { \
                     ^
caffe2/aten/src/ATen/TensorOperators.h:110:22: note: candidate function not viable: no known conversion from 'TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' to 'const at::Tensor' for 1st argument
caffe2/aten/src/ATen/TensorOperators.h:82:33: note: expanded from macro 'AT_FORALL_BINARY_OPS'
#define AT_FORALL_BINARY_OPS(_) \
                                ^
caffe2/aten/src/ATen/TensorOperators.h:99:22: note: expanded from macro 'DEFINE_OPERATOR'
static inline Tensor operator op(const Tensor & x, const Tensor & y) { \
                     ^
caffe2/aten/src/ATen/TensorOperators.h:110:22: note: candidate function not viable: no known conversion from 'TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' to 'c10::Scalar' for 1st argument
caffe2/aten/src/ATen/TensorOperators.h:82:33: note: expanded from macro 'AT_FORALL_BINARY_OPS'
#define AT_FORALL_BINARY_OPS(_) \
                                ^
caffe2/aten/src/ATen/TensorOperators.h:105:22: note: expanded from macro 'DEFINE_OPERATOR'
static inline Tensor operator op(Scalar x, const Tensor & y) { \
                     ^
In module 'third_party_libgcc_std' imported from caffe2/c10/core/Allocator.h:4:
third-party-buck/platform007/build/libgcc/include/c++/7.3.0/bits/stl_algobase.h:753:29: error: cannot increment value of type 'at::TensorAccessor<long, 1, DefaultPtrTraits, long>'
           __niter > 0; --__niter, ++__first)
                                   ^ ~~~~~~~
third-party-buck/platform007/build/libgcc/include/c++/7.3.0/bits/stl_algobase.h:789:23: note: in instantiation of function template specialization 'std::__fill_n_a<at::TensorAccessor<long, 1, DefaultPtrTraits, long>, long, int>' requested here
      return _OI(std::__fill_n_a(std::__niter_base(__first), __n, __value));
                      ^
caffe2/aten/src/ATen/native/SobolEngineOps.cpp:133:8: note: in instantiation of function template specialization 'std::fill_n<at::TensorAccessor<long, 1, DefaultPtrTraits, long>, long, int>' requested here
  std::fill_n(ss_a[0], MAXBIT, 1);
       ^
2 errors generated.
    When running <c++ preprocess_and_compile>.
    When building rule //caffe2/aten:ATen-cpu#compile-pic-SobolEngineOps.cpp.o6f3daec3,platform007-clang.

**Pull request closed without merging**